// @fusiongrid/datastore/src/toolbox/src/index.js
import convertKeysToLowerCase from './json-utils/convert-keys-lower-case';
import { mergeDeep, mergeArray, clone } from './json-utils/merge-deep';
import formatString from './string-utils/format-string';
import { captalizeFirstLetter } from './string-utils/string-case';
import EventManager from './managers/event-manager';
import syncScroll from './managers/scroll-manager';
import { getSpanOffsetHeight } from './units/dom-dimensions';
import { parseLengthUnit } from './units/parse-length';
import Logger from './managers/log-manager';
import { durationSecond, durationMinute, durationHour, durationDay, durationWeek, durationMonth, durationQuarter, durationYear } from './datetime/durations';
import { DatetimeUnits, Weekdays } from './datetime/datetime-enums';
import { isNumber, isNaturalNumber, isWholeNumber } from './type-utils';
import { numberComparator, stringComparator } from './comparators';
import uid, { numberIdentity } from './uid';


export { 
  Logger,
  convertKeysToLowerCase,
  mergeDeep,
  mergeArray,
  clone,
  formatString,
  captalizeFirstLetter,
  EventManager,
  parseLengthUnit,
  getSpanOffsetHeight,
  syncScroll,
  durationSecond, durationMinute, durationHour, durationDay, durationWeek, durationMonth, durationQuarter, durationYear,
  DatetimeUnits, Weekdays,
  isNumber, isNaturalNumber, isWholeNumber, numberComparator, stringComparator,
  uid, numberIdentity
};
