// fusiongrid/texts/exception/grid-core.ts
export const GridException = {
  parameterMissing: 'Unable to render grid. DOM Container or data is missing in grid initialization.',
  containerNotElement: 'Grid conatiner must be HTML element',
  dataNotDataTable: 'Data must be passed to grid as DataTable object',
  fieldNotFound: 'Except an HTML type column, all columns must define a field property. This denotes the column/field of data table with which this column is bound to.',
  fieldError: 'The field property of a columm must refer to an existing column of DataTable. Field: {0} is not found in DataTable',
  inlineChartTypeMismatch: '{0} is not a number type field and can not be used for inline charts.',
  noColumnFound: 'No columns to show. Column definition do not match with schema provided',
  minGreaterMaxWidth: 'Both minimum and maximum width value is ignored for column {0} as minimum value is greater than maximum value',
  viewportAdditionFail: 'Failed to add one or more viewports due lack of mandatory information(s)',
  invalidDefaultPageSize: 'Provided default page size is not valid, hence the first value from "options" is set as default page size',
  invalidPageSizeOptionsArr: 'All page size options in the provided array is non numeric or invalid',
  invalidPageSizeOptions: 'Options array for page size can only be a boolean value or numeric array',
  sizeColumnsToFitFailed: 'Columns can not be fit inside viewing area maintaining the minimum width of column.',
  sizeColumnsToContentFailedForColumn: 'Unable to apply fit to content behavior for invalid column indices.',
  invalidRowHeight: 'Value provied for attribute rowHeight is not a valid number',
  invalidHeaderRowHeight: 'Value provied for attribute headerRowHeight is not a valid number',
  inValidColumnIndex: 'Value provided as columnIndex is not a valid whole number',
  acceptsOnlyArrayAsInput: 'Invalid value provided, the API accepts only array as input',
  invalidLayoutInput: 'Invalid value provided, the API only accepts object with valid layout configurations or layout configuration in key value pair',
  invalidLayoutConfiguration: 'Invalid value provided, the API only accepts valid layout configuration in form of string or array of strings',
  invalidRowIndex: 'Invalid row index. Please try again with valid index'
};
